pymc3/step_methods/sgmcmc.py

```diff
@@ -0,0 +1,295 @@
+from collections import OrderedDict
+import warnings
+
+from .arraystep import Competence, ArrayStepShared
+from ..vartypes import continuous_types
+from ..model import modelcontext, inputvars
+import theano.tensor as tt
+from ..theanof import tt_rng, make_shared_replacements
+import theano
+import numpy as np
+
+__all__ = ['SGFS']
+
+EXPERIMENTAL_WARNING = "Warning: Stocastic Gradient based sampling methods are experimental step methods and not yet"\
+    " recommended for use in PyMC3!"
+
+def _value_error(cond, str):
+    """Throws ValueError if cond is False"""
+    if not cond:
+        raise ValueError(str)
+
+def _check_minibatches(minibatch_tensors, minibatches):
+    _value_error(isinstance(minibatch_tensors, list),
+                 'minibatch_tensors must be a list.')
+
+    _value_error(hasattr(minibatches, "__iter__"),
+                 'minibatches must be an iterator.')
+
+def prior_dlogp(vars, model, flat_view):
+    """Returns the gradient of the prior on the parameters as a vector of size D x 1"""
+    terms = tt.concatenate(
+        [theano.grad(var.logpt, var).flatten() for var in vars],
+        axis=0)
+    dlogp = theano.clone(terms, flat_view.replacements, strict=False)
+
+    return dlogp
+
+def elemwise_dlogL(vars, model, flat_view):
+    """
+    Returns Jacobian of the log likelihood for each training datum wrt vars
+    as a matrix of size N x D
+    """
+    # select one observed random variable 
+    obs_var = model.observed_RVs[0]
+    # tensor of shape (batch_size,)
+    logL = obs_var.logp_elemwiset.sum(axis=tuple(range(1, obs_var.logp_elemwiset.ndim)))
+    # calculate fisher information
+    terms = []
+    for var in vars:
+         output, _ =  theano.scan(lambda i, logX=logL, v=var: theano.grad(logX[i], v).flatten(),\
+                            sequences=[tt.arange(logL.shape[0])])
+         terms.append(output)
+    dlogL = theano.clone(tt.concatenate(terms, axis=1), flat_view.replacements, strict=False)
+    return dlogL
+
+
+class BaseStochasticGradient(ArrayStepShared):
+    R"""
+    BaseStochasticGradient Object
+    
+    For working with BaseStochasticGradient Object
+    we need to supply the probabilistic model 
+    (:code:`model`) with the data supplied to observed
+    variables of type `GeneratorOp`
+    
+    Parameters
+    -------
+    vars : list
+        List of variables for sampler
+    batch_size`: int
+        Batch Size for each step
+    total_size : int
+        Total size of the training data
+    step_size : float
+        Step size for the parameter update
+    model : PyMC Model
+        Optional model for sampling step. Defaults to None (taken from context)
+    random_seed : int
+        The seed to initialize the Random Stream
+    minibatches : iterator
```

Comment:
@twiecki @ferrine 

Okay I am running into two issues with the sgfs bayesian neural network notebook. 

1. the sampling speed drops by 50% when I use pm.Minibatch
2. the batch size is not configurable. My test and training datasets have different sizes, 50 and 500.

The sampling steps are also unusually large (high mixing) when I use pm.Minibatch instead of a simple training data iterator.

Let me know if you want me to create a separate issue for that after this PR. Otherwise we can discuss here, but it can get confusing.